romancal/resample/tests/test_resample_step.py

```diff
@@ -199,6 +203,74 @@ def test_load_custom_wcs_asdf_without_wcs_attribute(tmp_path):
         step._load_custom_wcs(str(file_path), (100, 100))
 
 
+def _assert_frame_equal(a, b):
+    """Copied from `gwcs`'s test_wcs.py"""
+    __tracebackhide__ = True
+
+    assert type(a) is type(b)
+
+    if a is None:
+        return
+
+    if not isinstance(a, cf.CoordinateFrame):
+        return a == b
+
+    assert a.name == b.name  # nosec
+    assert a.axes_order == b.axes_order  # nosec
+    assert a.axes_names == b.axes_names  # nosec
+    assert a.unit == b.unit  # nosec
+    assert a.reference_frame == b.reference_frame  # nosec
+
+
+def _assert_wcs_equal(a, b):
+    """Based on corresponding function from `gwcs`'s test_wcs.py"""
+    assert a.name == b.name  # nosec
+
+    assert a.pixel_shape == b.pixel_shape
+    assert a.array_shape == b.array_shape
+    if a.array_shape is not None:
+        assert a.array_shape == b.pixel_shape[::-1]
+
+    assert len(a.available_frames) == len(b.available_frames)  # nosec
+    for a_step, b_step in zip(a.pipeline, b.pipeline, strict=False):
+        _assert_frame_equal(a_step.frame, b_step.frame)
+        assert_model_equal(a_step.transform, b_step.transform)
+
+
+@pytest.mark.parametrize(
+    "wcs_type",
+    ["obj", "bytes"],
+)
+def test_load_custom_wcs_obj(exposure_1, wcs_type):
+    shape = (123, 456)
+    wcs_obj = create_wcs_object_without_distortion(
+        (10, 0),
+        (0.000031, 0.000031),
+        shape,
+    )
+    step = ResampleStep()
+
+    if wcs_type == "obj":
+        step.output_wcs = wcs_obj
+
+    elif wcs_type == "bytes":
+        # write to a byte stream:
+        bstream = BytesIO()
+        AsdfFile({"wcs": wcs_obj}).write_to(bstream)
+        bstream.seek(0)
+        step.output_wcs = bstream
+
+    result = step.process(ModelLibrary(exposure_1))
```

Comment:
```suggestion
    result = step.run(ModelLibrary(exposure_1))
```
The mosaic pipeline uses `run` (as recommended). If I change this to run the test fails with: `stpipe.config_parser.ValidationError: Config parameter 'output_wcs'`

Reply:
This PR addresses specific use case indicated in https://github.com/spacetelescope/romancal/issues/1404#issuecomment-2669420348.

Reply:
The comment mentions lines 133-135 in the mosaic pipeline:
https://github.com/spacetelescope/romancal/blob/d2ff8dee06348be51b9782a73cb1100e9cdfc3cc/romancal/pipeline/mosaic_pipeline.py#L133-L135
A bit further down where resample is called the pipeline uses `resample.run`:
https://github.com/spacetelescope/romancal/blob/d2ff8dee06348be51b9782a73cb1100e9cdfc3cc/romancal/pipeline/mosaic_pipeline.py#L148
The suggestion here is to update the test you've added to use `run` instead of `process`. When I tested this change locally it failed with the error mentioned above.

Reply:
The issue appears to be this line is stpipe:
https://github.com/spacetelescope/stpipe/blob/47c3867cdab9d8c9fb13150047df44cfbbd97f9e/src/stpipe/step.py#L443
which prints out the step parameters when the step is called in this test. Since `output_wcs` is defined as a string in the spec:
https://github.com/spacetelescope/romancal/blob/d2ff8dee06348be51b9782a73cb1100e9cdfc3cc/romancal/resample/resample_step.py#L68
this fails since the `get_pars` triggers validation against the spec.

When a step is part of a pipeline the parameters are printed at the start of `Pipeline.run` (which occurs before `Pipeline.process` where `resample.run` is called). This means that updating the test to use `run` fails (due to the parameters being printed after the `output_wcs` is assigned) but the mosaic pipeline usage prints the step specs prior to `output_wcs` being assigned and doesn't trigger the validation (running with an invalid spec).

Based on our previous conversation I was under the impression that this PR was going to take a different approach where the skycell wcs would be computed in resample (since all the relevant information is passed to resample via the association). That approach would not rely on being able to run an invalid step as long as it's made invalid during a pipeline run. Is that no longer the approach you're in favor of?

Reply:
https://github.com/spacetelescope/romancal/pull/1642 implements the approach discussed above where the skycell wcs is determined from the ModelLibrary association information within resample. This has a few benefits:
- allows the skycell wcs to be used in outlier detection with no code changes (addressing https://github.com/spacetelescope/romancal/issues/1615)
- simplifies the mosaic pipeline
- doesn't require providing invalid input

I also did some more testing with the approach in this PR (passing a `gwcs.WCS` instance to `output_wcs`). It would be possible to update the test to use `run` if the step spec was changed to:
```
output_wcs = pass(default='')  # Custom output WCS. 
```
which disables validation of `output_wcs`.